In [1]:
# =============================================================================
# STEP 1: LOAD PROCESSED DATASET
# =============================================================================

import pandas as pd
import numpy as np


# Load cleaned dataset
df = pd.read_csv(
    "../../../data/processed/heartdisease_cleaned.csv"
)


print("=" * 70)
print("PROCESSED DATASET")
print("=" * 70)

print(f"Rows    : {df.shape[0]}")
print(f"Columns : {df.shape[1]}")


display(df.head())

PROCESSED DATASET
Rows    : 303
Columns : 14


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63.0,1.0,1.0,145.0,233.0,1.0,2.0,150.0,0.0,2.3,3.0,0.0,6.0,0
1,67.0,1.0,4.0,160.0,286.0,0.0,2.0,108.0,1.0,1.5,2.0,3.0,3.0,1
2,67.0,1.0,4.0,120.0,229.0,0.0,2.0,129.0,1.0,2.6,2.0,2.0,7.0,1
3,37.0,1.0,3.0,130.0,250.0,0.0,0.0,187.0,0.0,3.5,3.0,0.0,3.0,0
4,41.0,0.0,2.0,130.0,204.0,0.0,2.0,172.0,0.0,1.4,1.0,0.0,3.0,0


In [2]:
# =============================================================================
# STEP 2: FEATURE AND TARGET SEPARATION
# =============================================================================

target_column = "target"


X = df.drop(
    target_column,
    axis=1
)


y = df[target_column]


print("=" * 70)
print("FEATURE MATRIX (X)")
print("=" * 70)

print(f"Rows    : {X.shape[0]}")
print(f"Columns : {X.shape[1]}")


print("\n" + "=" * 70)
print("TARGET VARIABLE (y)")
print("=" * 70)

print(f"Rows : {y.shape[0]}")


display(y.head())

FEATURE MATRIX (X)
Rows    : 303
Columns : 13

TARGET VARIABLE (y)
Rows : 303


0    0
1    1
2    1
3    0
4    0
Name: target, dtype: int64

In [3]:
# =============================================================================
# STEP 3: VERIFY TARGET DISTRIBUTION
# =============================================================================

print("=" * 70)
print("TARGET DISTRIBUTION")
print("=" * 70)


display(
    y.value_counts()
)


print("\nClass Percentage")


display(
    (y.value_counts(normalize=True) * 100)
    .round(2)
)

TARGET DISTRIBUTION


target
0    164
1    139
Name: count, dtype: int64


Class Percentage


target
0    54.13
1    45.87
Name: proportion, dtype: float64

In [4]:
# =============================================================================
# STEP 4: FEATURE TYPE CHECK
# =============================================================================

print("=" * 70)
print("FEATURE DATA TYPES")
print("=" * 70)


display(
    X.dtypes.to_frame(
        name="Data Type"
    )
)

FEATURE DATA TYPES


,Data Type
age,float64
sex,float64
cp,float64
trestbps,float64
chol,float64
fbs,float64
restecg,float64
thalach,float64
exang,float64
oldpeak,float64


In [5]:
# =============================================================================
# STEP 5: MISSING VALUE CHECK
# =============================================================================

print("=" * 70)
print("MISSING VALUES")
print("=" * 70)


missing = pd.DataFrame({
    "Missing Count": X.isnull().sum(),
    "Missing Percentage (%)": (
        X.isnull().mean() * 100
    ).round(2)
})


display(
    missing[missing["Missing Count"] > 0]
)

MISSING VALUES


,Missing Count,Missing Percentage (%)
ca,4,1.32
thal,2,0.66


In [6]:
# =============================================================================
# STEP 6: MISSING VALUE IMPUTATION
# =============================================================================

from sklearn.impute import SimpleImputer


imputer = SimpleImputer(
    strategy="median"
)


X_imputed = pd.DataFrame(
    imputer.fit_transform(X),
    columns=X.columns,
    index=X.index
)


print("=" * 70)
print("MISSING VALUE IMPUTATION")
print("=" * 70)

print("Imputation completed successfully")


print("\nRemaining Missing Values:")
print(X_imputed.isnull().sum().sum())

MISSING VALUE IMPUTATION
Imputation completed successfully

Remaining Missing Values:
0


In [7]:
# =============================================================================
# STEP 7: TRAIN-TEST SPLIT
# =============================================================================

from sklearn.model_selection import train_test_split


X_train, X_test, y_train, y_test = train_test_split(
    X_imputed,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)


print("=" * 70)
print("TRAIN-TEST SPLIT")
print("=" * 70)


print("Training Data")
print("-" * 70)

print(f"X_train shape : {X_train.shape}")
print(f"y_train shape : {y_train.shape}")


print("\nTesting Data")
print("-" * 70)

print(f"X_test shape  : {X_test.shape}")
print(f"y_test shape  : {y_test.shape}")

TRAIN-TEST SPLIT
Training Data
----------------------------------------------------------------------
X_train shape : (242, 13)
y_train shape : (242,)

Testing Data
----------------------------------------------------------------------
X_test shape  : (61, 13)
y_test shape  : (61,)


In [8]:
# =============================================================================
# STEP 8: CLASS DISTRIBUTION AFTER SPLIT
# =============================================================================

print("=" * 70)
print("TRAINING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_train.value_counts(normalize=True) * 100)
    .round(2)
)


print("\n" + "=" * 70)
print("TESTING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_test.value_counts(normalize=True) * 100)
    .round(2)
)

TRAINING TARGET DISTRIBUTION


target
0    54.13
1    45.87
Name: proportion, dtype: float64


TESTING TARGET DISTRIBUTION


target
0    54.1
1    45.9
Name: proportion, dtype: float64

In [9]:
# =============================================================================
# STEP 9: FEATURE SCALING
# =============================================================================

from sklearn.preprocessing import StandardScaler


print("=" * 70)
print("FEATURE SCALING USING STANDARD SCALER")
print("=" * 70)


scaler = StandardScaler()


# Fit ONLY on training data
X_train_scaled = scaler.fit_transform(
    X_train
)


# Transform test data
X_test_scaled = scaler.transform(
    X_test
)


print("Scaling completed successfully")


print("\n" + "=" * 70)
print("SHAPE AFTER SCALING")
print("=" * 70)


print(
    f"X_train_scaled shape : {X_train_scaled.shape}"
)

print(
    f"X_test_scaled shape  : {X_test_scaled.shape}"
)

FEATURE SCALING USING STANDARD SCALER
Scaling completed successfully

SHAPE AFTER SCALING
X_train_scaled shape : (242, 13)
X_test_scaled shape  : (61, 13)


In [10]:
# =============================================================================
# STEP 10: CONVERT SCALED DATA BACK TO DATAFRAME
# =============================================================================


X_train_scaled = pd.DataFrame(
    X_train_scaled,
    columns=X_train.columns,
    index=X_train.index
)


X_test_scaled = pd.DataFrame(
    X_test_scaled,
    columns=X_test.columns,
    index=X_test.index
)


print("=" * 70)
print("SCALED TRAIN DATA PREVIEW")
print("=" * 70)


display(
    X_train_scaled.head()
)

SCALED TRAIN DATA PREVIEW


,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal
180,-0.729485,0.68313,0.870169,-0.395692,0.458139,-0.411196,1.022996,0.708371,-0.696177,-0.445445,0.675060,-0.689715,1.179973
208,0.050166,0.68313,-1.184278,-0.054513,0.230598,-0.411196,-0.981579,0.222495,-0.696177,-0.891627,-0.958585,-0.689715,-0.878070
167,-0.061212,-1.46385,-1.184278,0.059213,0.723605,2.431930,1.022996,0.399178,1.436416,-0.891627,-0.958585,0.445734,-0.878070
105,-0.061212,0.68313,-1.184278,-1.305501,1.121803,-0.411196,-0.981579,0.266666,-0.696177,-0.891627,-0.958585,-0.689715,1.179973
297,0.272924,-1.46385,0.870169,0.514117,-0.167601,-0.411196,-0.981579,-1.190962,1.436416,-0.713154,0.675060,-0.689715,1.179973


In [11]:
# =============================================================================
# STEP 11: SCALING VERIFICATION
# =============================================================================

print("=" * 70)
print("MEAN AFTER SCALING")
print("=" * 70)


display(
    X_train_scaled.mean().round(2)
)


print("\n" + "=" * 70)
print("STANDARD DEVIATION AFTER SCALING")
print("=" * 70)


display(
    X_train_scaled.std().round(2)
)

MEAN AFTER SCALING


age        -0.0
sex         0.0
cp          0.0
trestbps   -0.0
chol        0.0
fbs        -0.0
restecg     0.0
thalach    -0.0
exang       0.0
oldpeak     0.0
slope      -0.0
ca          0.0
thal        0.0
dtype: float64


STANDARD DEVIATION AFTER SCALING


age         1.0
sex         1.0
cp          1.0
trestbps    1.0
chol        1.0
fbs         1.0
restecg     1.0
thalach     1.0
exang       1.0
oldpeak     1.0
slope       1.0
ca          1.0
thal        1.0
dtype: float64

# LASSO

## 1. Baseline

In [12]:
# =============================================================================
# LASSO BASELINE FEATURE SELECTION (TOP-K = 5,10,15,20)
# =============================================================================

from sklearn.linear_model import LogisticRegression
import pandas as pd
import numpy as np


# Fit LASSO once with all features
# Logistic Regression with L1 penalty = LASSO-style feature selection

lasso_selector = LogisticRegression(
    penalty="l1",
    solver="liblinear",
    C=0.1,
    max_iter=2000,
    class_weight="balanced",
    random_state=42
)


lasso_selector.fit(
    X_train_scaled,
    y_train
)


# =============================================================================
# Extract LASSO coefficients
# =============================================================================

lasso_coefficients = lasso_selector.coef_[0]


# Use absolute coefficient magnitude as feature importance
lasso_scores = pd.DataFrame({

    "Feature": X_train_scaled.columns,

    "LASSO Coefficient": lasso_coefficients,

    "LASSO Score": np.abs(lasso_coefficients)

})


# Sort by absolute coefficient magnitude
lasso_scores = lasso_scores.sort_values(
    by="LASSO Score",
    ascending=False
).reset_index(drop=True)


# =============================================================================
# Top-K values
# =============================================================================

k_values = [5, 8, 10, 13]

lasso_results = {}


for k in k_values:

    top_features = (
        lasso_scores
        .head(k)["Feature"]
        .tolist()
    )


    # Boolean mask for selecting columns
    mask = X_train_scaled.columns.isin(
        top_features
    )


    X_train_selected = X_train_scaled.loc[
        :, mask
    ]

    X_test_selected = X_test_scaled.loc[
        :, mask
    ]


    lasso_results[k] = {

        "features": top_features,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "scores": lasso_scores.head(k)

    }


    print("=" * 70)
    print(f"TOP {k} LASSO FEATURES")
    print("=" * 70)

    print(top_features)

    print("\nScores:")
    print(lasso_scores.head(k))

TOP 5 LASSO FEATURES
['ca', 'thal', 'cp', 'exang', 'sex']

Scores:
  Feature  LASSO Coefficient  LASSO Score
0      ca           0.643613     0.643613
1    thal           0.643237     0.643237
2      cp           0.280976     0.280976
3   exang           0.259048     0.259048
4     sex           0.227397     0.227397
TOP 8 LASSO FEATURES
['ca', 'thal', 'cp', 'exang', 'sex', 'thalach', 'slope', 'oldpeak']

Scores:
   Feature  LASSO Coefficient  LASSO Score
0       ca           0.643613     0.643613
1     thal           0.643237     0.643237
2       cp           0.280976     0.280976
3    exang           0.259048     0.259048
4      sex           0.227397     0.227397
5  thalach          -0.177869     0.177869
6    slope           0.141674     0.141674
7  oldpeak           0.050402     0.050402
TOP 10 LASSO FEATURES
['ca', 'thal', 'cp', 'exang', 'sex', 'thalach', 'slope', 'oldpeak', 'restecg', 'chol']

Scores:
   Feature  LASSO Coefficient  LASSO Score
0       ca           0.643613     0

In [13]:
# =============================================================================
# BASELINE MODELS
# =============================================================================

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier


models = {

    "Logistic Regression": LogisticRegression(
        max_iter=1000,
        class_weight="balanced",
        random_state=42
    ),

    "Random Forest": RandomForestClassifier(
        n_estimators=100,
        class_weight="balanced",
        random_state=42
    ),

    "XGBoost": XGBClassifier(
        random_state=42,
        eval_metric="logloss",
        scale_pos_weight=(
            (y_train == 0).sum() /
            (y_train == 1).sum()
        )
    )
}

In [14]:
# =============================================================================
# MODEL EVALUATION — LASSO
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)

results_lasso = []


for k in k_values:

    Xtr = lasso_results[k]["X_train"]

    Xts = lasso_results[k]["X_test"]

    features = ", ".join(
        lasso_results[k]["features"]
    )


    for model_name, model in models.items():

        model.fit(
            Xtr,
            y_train
        )


        y_pred = model.predict(
            Xts
        )


        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        results_lasso.append({

            "Method": "LASSO",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred
            ),

            "Recall": recall_score(
                y_test,
                y_pred
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )

        })


lasso_evaluation = pd.DataFrame(
    results_lasso
)

display(lasso_evaluation)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,LASSO,5,Logistic Regression,"ca, thal, cp, exang, sex",0.819672,0.774194,0.857143,0.813559,0.898268
1,LASSO,5,Random Forest,"ca, thal, cp, exang, sex",0.819672,0.793103,0.821429,0.807018,0.904762
2,LASSO,5,XGBoost,"ca, thal, cp, exang, sex",0.836066,0.800000,0.857143,0.827586,0.924242
3,LASSO,8,Logistic Regression,"ca, thal, cp, exang, sex, thalach, slope, oldpeak",0.852459,0.806452,0.892857,0.847458,0.931818
4,LASSO,8,Random Forest,"ca, thal, cp, exang, sex, thalach, slope, oldpeak",0.836066,0.800000,0.857143,0.827586,0.943182
5,LASSO,8,XGBoost,"ca, thal, cp, exang, sex, thalach, slope, oldpeak",0.819672,0.774194,0.857143,0.813559,0.930736
6,LASSO,10,Logistic Regression,"ca, thal, cp, exang, sex, thalach, slope, oldp...",0.852459,0.787879,0.928571,0.852459,0.932900
7,LASSO,10,Random Forest,"ca, thal, cp, exang, sex, thalach, slope, oldp...",0.885246,0.838710,0.928571,0.881356,0.941558
8,LASSO,10,XGBoost,"ca, thal, cp, exang, sex, thalach, slope, oldp...",0.836066,0.781250,0.892857,0.833333,0.913420
9,LASSO,13,Logistic Regression,"ca, thal, cp, exang, sex, thalach, slope, oldp...",0.868852,0.812500,0.928571,0.866667,0.950216


In [15]:
lasso_evaluation.to_csv(
    "../../../results/heart_disease/baseline/lasso_baseline_results.csv",
    index=False
)

## 2. DODA

In [16]:
import sys
import subprocess

subprocess.check_call([
    sys.executable,
    "-m",
    "pip",
    "install",
    "--upgrade",
    "--force-reinstall",
    "--no-cache-dir",
    "git+https://github.com/anandha-3679/DODA.git"
])

0

In [17]:
import sys
import doda
from doda import DODASelector
import inspect

print("Python:", sys.executable)
print("DODA:", doda.__file__)
print("Selector:", inspect.getfile(DODASelector))

Python: c:\Users\Anandha Lakshmi\Downloads\Anaconda\python.exe
DODA: c:\Users\Anandha Lakshmi\Downloads\Anaconda\Lib\site-packages\doda\__init__.py
Selector: c:\Users\Anandha Lakshmi\Downloads\Anaconda\Lib\site-packages\doda\selector.py


In [18]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

from doda import DODASelector

from doda.adapters import SklearnAdapter

from doda.knowledge.providers import JSONProvider

from doda.fusion import RankFusion

In [19]:
# =============================================================================
# LASSO + DODA FEATURE SELECTION
# =============================================================================

import pandas as pd
from sklearn.linear_model import LogisticRegression


# -------------------------------------------------------------------------
# Top-K values
# -------------------------------------------------------------------------

k_values = [5, 8, 10, 13]


# -------------------------------------------------------------------------
# Clinical knowledge provider
# -------------------------------------------------------------------------

provider = JSONProvider(
    "heartdisease_weights.json"
)


# -------------------------------------------------------------------------
# Fusion method
# -------------------------------------------------------------------------

fusion = RankFusion()


# -------------------------------------------------------------------------
# Store DODA results
# -------------------------------------------------------------------------

results = {}


# =============================================================================
# RUN LASSO + DODA FOR EACH TOP-K
# =============================================================================

for k in k_values:

    print("\n" + "=" * 70)
    print(f"LASSO + DODA — TOP {k} FEATURES")
    print("=" * 70)


    # ---------------------------------------------------------------------
    # LASSO operator
    # ---------------------------------------------------------------------

    lasso = SklearnAdapter(

        LogisticRegression(
            penalty="l1",
            solver="liblinear",
            C=0.1,
            max_iter=2000,
            class_weight="balanced",
            random_state=42
        )

    )


    # ---------------------------------------------------------------------
    # DODA selector
    # ---------------------------------------------------------------------

    selector = DODASelector(

        operators=[lasso],

        provider=provider,

        fusion=fusion,

        top_k=k
    )


    # ---------------------------------------------------------------------
    # Fit DODA on TRAINING data
    # ---------------------------------------------------------------------

    X_train_selected = selector.fit_transform(

        X_train_scaled,

        y_train

    )


    # ---------------------------------------------------------------------
    # Transform TEST data
    # ---------------------------------------------------------------------

    X_test_selected = selector.transform(

        X_test_scaled

    )


    # ---------------------------------------------------------------------
    # Selected features
    # ---------------------------------------------------------------------

    selected_features = (
        selector.get_selected_features()
    )


    print("\nSelected Features:")
    print(selected_features)


    # ---------------------------------------------------------------------
    # Store everything needed for later evaluation
    # ---------------------------------------------------------------------

    results[k] = {

        "selector": selector,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "features": selected_features,

        "raw_math_scores":
            selector.raw_math_scores_,

        "math_scores":
            selector.math_scores_,

        "clinical_weights":
            selector.clinical_weights_,

        "final_scores":
            selector.final_scores_
    }


print("\n" + "=" * 70)
print("LASSO + DODA FEATURE SELECTION COMPLETED")
print("=" * 70)


LASSO + DODA — TOP 5 FEATURES
Provider inside knowledge engine: <doda.knowledge.providers.json_provider.JSONProvider object at 0x00000207B62C3A10>
DODA Pipeline Started
Running: LogisticRegression
DODA feature order:
['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak']
Number of features: 13
Number of coefficients: 13
Operator Scores:
{'LogisticRegression_1': {'age': np.float64(0.0), 'sex': np.float64(0.2273969834700217), 'cp': np.float64(0.28097562421045086), 'trestbps': np.float64(0.0), 'chol': np.float64(0.0), 'fbs': np.float64(0.0), 'restecg': np.float64(0.011081865387417202), 'thalach': np.float64(0.17786897896866977), 'exang': np.float64(0.2590475736510421), 'oldpeak': np.float64(0.05040193546062819), 'slope': np.float64(0.14167422735296537), 'ca': np.float64(0.6436128225470914), 'thal': np.float64(0.6432367714139653)}}
Resolving scores from: 1 operators

Raw Mathematical Scores
{'age': np.float64(0.0), 'sex': np.float64(0.2273969834700217), 

In [20]:
# =============================================================================
# INSPECT MATHEMATICAL SCORES
# =============================================================================

k = 10

selector = results[k]["selector"]


# -------------------------------------------------------------------------
# Raw mathematical scores
# -------------------------------------------------------------------------

raw_math_scores = pd.DataFrame(
    selector.raw_math_scores_.items(),
    columns=[
        "Feature",
        "Raw Math Score"
    ]
)


# -------------------------------------------------------------------------
# Normalized mathematical scores
# -------------------------------------------------------------------------

normalized_math_scores = pd.DataFrame(
    selector.math_scores_.items(),
    columns=[
        "Feature",
        "Normalized Math Score"
    ]
)


# -------------------------------------------------------------------------
# Combine both
# -------------------------------------------------------------------------

math_scores = raw_math_scores.merge(
    normalized_math_scores,
    on="Feature"
)


math_scores = math_scores.sort_values(
    by="Normalized Math Score",
    ascending=False
)


print("=" * 70)
print("MATHEMATICAL SCORES")
print("=" * 70)

display(math_scores)

MATHEMATICAL SCORES


,Feature,Raw Math Score,Normalized Math Score
11,ca,0.643613,1.000000
12,thal,0.643237,0.999416
2,cp,0.280976,0.436560
8,exang,0.259048,0.402490
1,sex,0.227397,0.353313
7,thalach,0.177869,0.276360
10,slope,0.141674,0.220123
9,oldpeak,0.050402,0.078311
6,restecg,0.011082,0.017218
4,chol,0.000000,0.000000


In [21]:
# =============================================================================
# INSPECT CLINICAL WEIGHTS
# =============================================================================

clinical_weights = pd.DataFrame(
    selector.clinical_weights_.items(),
    columns=[
        "Feature",
        "Clinical Weight"
    ]
)


clinical_weights = clinical_weights.sort_values(
    by="Clinical Weight",
    ascending=False
)


print("=" * 70)
print("CLINICAL WEIGHTS")
print("=" * 70)

display(clinical_weights)

CLINICAL WEIGHTS


,Feature,Clinical Weight
2,cp,1.0
9,oldpeak,1.0
11,ca,1.0
12,thal,1.0
8,exang,1.0
10,slope,0.9
0,age,0.9
3,trestbps,0.9
4,chol,0.8
7,thalach,0.8


In [22]:
# =============================================================================
# INSPECT FINAL RANK FUSION SCORES
# =============================================================================

final_scores = pd.DataFrame(
    selector.final_scores_.items(),
    columns=[
        "Feature",
        "Final Rank Fusion Score"
    ]
)


final_scores = final_scores.sort_values(
    by="Final Rank Fusion Score",
    ascending=False
)


print("=" * 70)
print("FINAL RANK FUSION SCORES")
print("=" * 70)

display(final_scores)

FINAL RANK FUSION SCORES


,Feature,Final Rank Fusion Score
11,ca,0.032787
12,thal,0.032522
2,cp,0.032266
8,exang,0.032018
9,oldpeak,0.031099
10,slope,0.031054
7,thalach,0.031025
1,sex,0.031010
0,age,0.030415
3,trestbps,0.030415


In [23]:
# =============================================================================
# SELECTED FEATURES
# =============================================================================

selector.get_selected_features()

['ca',
 'thal',
 'cp',
 'exang',
 'oldpeak',
 'slope',
 'thalach',
 'sex',
 'age',
 'trestbps']

In [24]:
# =============================================================================
# COMPARE RANKINGS
# =============================================================================

comparison = selector.compare_scores()

display(comparison)


MATHEMATICAL vs CLINICAL vs RANKFUSION

Fusion Method: RankFusion

Top Features After Fusion:
    Feature  Math Rank  Clinical Rank  Final Rank  Rank Change
0        ca          1              4           1            0
1      thal          2              5           2            0
2        cp          3              1           3            0
3     exang          4              2           4            0
4   oldpeak          8              3           5            3
5     slope          7              8           6            1
6   thalach          6             11           7           -1
7       sex          5             12           8           -3
8       age         10              6           9            1
9  trestbps         11              7          10            1


,Feature,Math Rank,Normalized Math Score,Clinical Rank,Clinical Weight,Final Rank,Final Score,Rank Change
0,ca,1,1.0000,4,1.0,1,0.0328,0
1,thal,2,0.9994,5,1.0,2,0.0325,0
2,cp,3,0.4366,1,1.0,3,0.0323,0
3,exang,4,0.4025,2,1.0,4,0.0320,0
4,oldpeak,8,0.0783,3,1.0,5,0.0311,3
5,slope,7,0.2201,8,0.9,6,0.0311,1
6,thalach,6,0.2764,11,0.8,7,0.0310,-1
7,sex,5,0.3533,12,0.6,8,0.0310,-3
8,age,10,0.0000,6,0.9,9,0.0304,1
9,trestbps,11,0.0000,7,0.9,10,0.0304,1


In [25]:
# =============================================================================
#  MODELS
# =============================================================================

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier


models = {

    "Logistic Regression":
        LogisticRegression(
            max_iter=1000,
            random_state=42
        ),


    "Random Forest":
        RandomForestClassifier(
            n_estimators=200,
            random_state=42
        ),


    "XGBoost":
        XGBClassifier(
            n_estimators=200,
            random_state=42,
            eval_metric="logloss"
        )
}

In [26]:
# =============================================================================
# RANK FUSION DODA MODEL EVALUATION
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)


doda_results = []


for k in k_values:

    Xtr = results[k]["X_train"]

    Xts = results[k]["X_test"]

    features = ", ".join(
        results[k]["features"]
    )


    for model_name, model in models.items():

        model.fit(
            Xtr,
            y_train
        )


        y_pred = model.predict(
            Xts
        )


        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        doda_results.append({

            "Method": "LASSO + Rank Fusion DODA",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "Recall": recall_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )
        })


doda_results_df = pd.DataFrame(
    doda_results
)


display(doda_results_df)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,LASSO + Rank Fusion DODA,5,Logistic Regression,"ca, thal, cp, exang, oldpeak",0.852459,0.806452,0.892857,0.847458,0.940476
1,LASSO + Rank Fusion DODA,5,Random Forest,"ca, thal, cp, exang, oldpeak",0.721311,0.666667,0.785714,0.721311,0.879870
2,LASSO + Rank Fusion DODA,5,XGBoost,"ca, thal, cp, exang, oldpeak",0.786885,0.758621,0.785714,0.771930,0.879329
3,LASSO + Rank Fusion DODA,8,Logistic Regression,"ca, thal, cp, exang, oldpeak, slope, thalach, sex",0.852459,0.806452,0.892857,0.847458,0.935065
4,LASSO + Rank Fusion DODA,8,Random Forest,"ca, thal, cp, exang, oldpeak, slope, thalach, sex",0.836066,0.764706,0.928571,0.838710,0.938853
5,LASSO + Rank Fusion DODA,8,XGBoost,"ca, thal, cp, exang, oldpeak, slope, thalach, sex",0.852459,0.827586,0.857143,0.842105,0.925325
6,LASSO + Rank Fusion DODA,10,Logistic Regression,"ca, thal, cp, exang, oldpeak, slope, thalach, ...",0.868852,0.812500,0.928571,0.866667,0.944805
7,LASSO + Rank Fusion DODA,10,Random Forest,"ca, thal, cp, exang, oldpeak, slope, thalach, ...",0.868852,0.812500,0.928571,0.866667,0.948052
8,LASSO + Rank Fusion DODA,10,XGBoost,"ca, thal, cp, exang, oldpeak, slope, thalach, ...",0.852459,0.771429,0.964286,0.857143,0.941558
9,LASSO + Rank Fusion DODA,13,Logistic Regression,"ca, thal, cp, exang, oldpeak, slope, thalach, ...",0.868852,0.812500,0.928571,0.866667,0.951299


In [27]:
# =============================================================================
# SAVE DODA RESULTS
# =============================================================================


doda_results_df.to_csv(
    "../../../results/heart_disease/doda/lasso_doda_results.csv",
    index=False
)

In [28]:
# =============================================================================
# COMBINED COMPARISON
# =============================================================================


comparison_df = pd.concat(
    [
        lasso_evaluation,
        doda_results_df
    ],
    ignore_index=True
)


display(
    comparison_df
)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,LASSO,5,Logistic Regression,"ca, thal, cp, exang, sex",0.819672,0.774194,0.857143,0.813559,0.898268
1,LASSO,5,Random Forest,"ca, thal, cp, exang, sex",0.819672,0.793103,0.821429,0.807018,0.904762
2,LASSO,5,XGBoost,"ca, thal, cp, exang, sex",0.836066,0.800000,0.857143,0.827586,0.924242
3,LASSO,8,Logistic Regression,"ca, thal, cp, exang, sex, thalach, slope, oldpeak",0.852459,0.806452,0.892857,0.847458,0.931818
4,LASSO,8,Random Forest,"ca, thal, cp, exang, sex, thalach, slope, oldpeak",0.836066,0.800000,0.857143,0.827586,0.943182
5,LASSO,8,XGBoost,"ca, thal, cp, exang, sex, thalach, slope, oldpeak",0.819672,0.774194,0.857143,0.813559,0.930736
6,LASSO,10,Logistic Regression,"ca, thal, cp, exang, sex, thalach, slope, oldp...",0.852459,0.787879,0.928571,0.852459,0.932900
7,LASSO,10,Random Forest,"ca, thal, cp, exang, sex, thalach, slope, oldp...",0.885246,0.838710,0.928571,0.881356,0.941558
8,LASSO,10,XGBoost,"ca, thal, cp, exang, sex, thalach, slope, oldp...",0.836066,0.781250,0.892857,0.833333,0.913420
9,LASSO,13,Logistic Regression,"ca, thal, cp, exang, sex, thalach, slope, oldp...",0.868852,0.812500,0.928571,0.866667,0.950216


In [29]:
# ============================================================
# DODA STABILITY ANALYSIS — ALL TOP-K VALUES
# Heart Disease Dataset
# ============================================================

import pandas as pd
import numpy as np

from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.impute import SimpleImputer
from sklearn.feature_selection import SelectKBest, f_classif

from doda import DODASelector
from doda.adapters.sklearn_adapter import SklearnAdapter
from doda.knowledge import JSONProvider
from doda.fusion import RankFusion


# ============================================================
# CONFIGURATION
# ============================================================

TOP_K_VALUES = [5, 8, 10, 13]

N_SPLITS = 5
N_REPEATS = 5
RANDOM_STATE = 42

WEIGHTS_FILE = "heartdisease_weights.json"


# ============================================================
# DATA
# ============================================================

# X and y should already contain the original Heart Disease data
# BEFORE imputation/scaling.

# Example:
# X = df.drop(columns=["target"])
# y = df["target"]


# ============================================================
# REPEATED STRATIFIED CV
# ============================================================

cv = RepeatedStratifiedKFold(
    n_splits=N_SPLITS,
    n_repeats=N_REPEATS,
    random_state=RANDOM_STATE
)


# ============================================================
# STORAGE
# ============================================================

all_runs = []
all_jaccard = []
all_frequency = []


# ============================================================
# LOOP THROUGH ALL TOP-K VALUES
# ============================================================

for TOP_K in TOP_K_VALUES:

    print("\n" + "=" * 70)
    print(f"DODA STABILITY ANALYSIS — TOP-{TOP_K}")
    print("=" * 70)

    selected_feature_sets = []

    run_number = 0

    # --------------------------------------------------------
    # 25 repeated CV runs
    # --------------------------------------------------------

    for train_idx, val_idx in cv.split(X, y):

        run_number += 1

        X_train = X.iloc[train_idx].copy()
        y_train = y.iloc[train_idx].copy()

        X_val = X.iloc[val_idx].copy()
        y_val = y.iloc[val_idx].copy()

        # ----------------------------------------------------
        # Imputation fitted ONLY on training data
        # ----------------------------------------------------

        imputer = SimpleImputer(strategy="median")

        X_train_imp = pd.DataFrame(
            imputer.fit_transform(X_train),
            columns=X_train.columns,
            index=X_train.index
        )

        X_val_imp = pd.DataFrame(
            imputer.transform(X_val),
            columns=X_val.columns,
            index=X_val.index
        )

        # ----------------------------------------------------
        # Lasso operator
        #
        # k="all" because DODA itself performs Top-K ranking
        # ----------------------------------------------------

        lasso_operator = SklearnAdapter(
            SelectKBest(
                score_func=f_classif,
                k="all"
            )
        )

        # ----------------------------------------------------
        # Clinical knowledge provider
        # ----------------------------------------------------

        provider = JSONProvider(WEIGHTS_FILE)

        # ----------------------------------------------------
        # DODA selector
        # ----------------------------------------------------

        doda_selector = DODASelector(
            operators=[lasso_operator],
            provider=provider,
            fusion=RankFusion(),
            top_k=TOP_K
        )

        # ----------------------------------------------------
        # Fit ONLY on training fold
        # ----------------------------------------------------

        doda_selector.fit(
            X_train_imp,
            y_train
        )

        # ----------------------------------------------------
        # Selected features
        # ----------------------------------------------------

        selected_features = list(
            doda_selector.get_selected_features()
        )

        selected_feature_sets.append(set(selected_features))

        all_runs.append({
            "Top_K": TOP_K,
            "Run": run_number,
            "Selected_Features": selected_features
        })

        print(
            f"Run {run_number:02d}: "
            f"{selected_features}"
        )


    # ========================================================
    # JACCARD SIMILARITY
    # ========================================================

    jaccard_scores = []

    for i in range(len(selected_feature_sets)):

        for j in range(i + 1, len(selected_feature_sets)):

            set_a = selected_feature_sets[i]
            set_b = selected_feature_sets[j]

            intersection = len(set_a.intersection(set_b))
            union = len(set_a.union(set_b))

            jaccard = intersection / union

            jaccard_scores.append(jaccard)

            all_jaccard.append({
                "Top_K": TOP_K,
                "Run_A": i + 1,
                "Run_B": j + 1,
                "Jaccard_Similarity": jaccard
            })


    # ========================================================
    # STABILITY SUMMARY
    # ========================================================

    mean_jaccard = np.mean(jaccard_scores)
    std_jaccard = np.std(jaccard_scores)
    min_jaccard = np.min(jaccard_scores)
    max_jaccard = np.max(jaccard_scores)

    print("\n" + "-" * 70)
    print(f"TOP-{TOP_K} STABILITY SUMMARY")
    print("-" * 70)

    print(f"Number of runs:          {len(selected_feature_sets)}")
    print(f"Number of comparisons:   {len(jaccard_scores)}")
    print(f"Mean Jaccard Similarity: {mean_jaccard:.4f}")
    print(f"Standard Deviation:      {std_jaccard:.4f}")
    print(f"Minimum Jaccard:         {min_jaccard:.4f}")
    print(f"Maximum Jaccard:         {max_jaccard:.4f}")


    # ========================================================
    # FEATURE SELECTION FREQUENCY
    # ========================================================

    feature_counts = {}

    for feature_set in selected_feature_sets:

        for feature in feature_set:

            feature_counts[feature] = (
                feature_counts.get(feature, 0) + 1
            )

    for feature, count in feature_counts.items():

        all_frequency.append({
            "Top_K": TOP_K,
            "Feature": feature,
            "Selection_Count": count,
            "Selection_Frequency": count / len(selected_feature_sets)
        })

    frequency_df = pd.DataFrame(
        [
            {
                "Feature": feature,
                "Selection_Count": count,
                "Selection_Frequency": count / len(selected_feature_sets)
            }
            for feature, count in feature_counts.items()
        ]
    ).sort_values(
        "Selection_Frequency",
        ascending=False
    )

    print("\nFEATURE SELECTION FREQUENCY")

    print(frequency_df.to_string(index=False))


# ============================================================
# CREATE DATAFRAMES
# ============================================================

runs_df = pd.DataFrame(all_runs)

jaccard_df = pd.DataFrame(all_jaccard)

frequency_df = pd.DataFrame(all_frequency)


# ============================================================
# SUMMARY TABLE
# ============================================================

summary_rows = []

for TOP_K in TOP_K_VALUES:

    subset = jaccard_df[
        jaccard_df["Top_K"] == TOP_K
    ]["Jaccard_Similarity"]

    summary_rows.append({
        "Top_K": TOP_K,
        "Mean_Jaccard": subset.mean(),
        "Std_Jaccard": subset.std(),
        "Min_Jaccard": subset.min(),
        "Max_Jaccard": subset.max()
    })


summary_df = pd.DataFrame(summary_rows)


# ============================================================
# DISPLAY FINAL SUMMARY
# ============================================================

print("\n" + "=" * 70)
print("DODA STABILITY — ALL TOP-K VALUES")
print("=" * 70)

print(
    summary_df.to_string(
        index=False,
        float_format=lambda x: f"{x:.4f}"
    )
)


# ============================================================
# SAVE RESULTS
# ============================================================

runs_df.to_csv(
    "../../../results/heart_disease/lasso_doda_stability_all_topk_runs.csv",
    index=False
)

jaccard_df.to_csv(
    "../../../results/heart_disease/lasso_doda_stability_all_topk_jaccard.csv",
    index=False
)

frequency_df.to_csv(
    "../../../results/heart_disease/lasso_doda_stability_all_topk_frequency.csv",
    index=False
)

summary_df.to_csv(
    "../../../results/heart_disease/lasso_doda_stability_all_topk_summary.csv",
    index=False
)


print("\nFiles saved:")
print("1. lasso_doda_stability_all_topk_runs.csv")
print("2. lasso_doda_stability_all_topk_jaccard.csv")
print("3. lasso_doda_stability_all_topk_frequency.csv")
print("4. lasso_doda_stability_all_topk_summary.csv")


DODA STABILITY ANALYSIS — TOP-5
Provider inside knowledge engine: <doda.knowledge.providers.json_provider.JSONProvider object at 0x00000207B7121A90>
DODA Pipeline Started
Running: SelectKBest
DODA feature order:
['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak']
Number of features: 13
Number of scores: 13
Operator Scores:
{'SelectKBest_1': {'age': np.float64(14.092089497961574), 'sex': np.float64(21.564202094745454), 'cp': np.float64(42.239456661366745), 'trestbps': np.float64(7.608511654167631), 'chol': np.float64(2.288972965863027), 'fbs': np.float64(0.11952693399811455), 'restecg': np.float64(6.906150433507175), 'thalach': np.float64(42.05786643867364), 'exang': np.float64(60.11520349331092), 'oldpeak': np.float64(53.44021864765956), 'slope': np.float64(27.81521670536931), 'ca': np.float64(57.94607446107259), 'thal': np.float64(82.69642608429048)}}
Resolving scores from: 1 operators

Raw Mathematical Scores
{'age': np.float64(14.09208949796157

In [30]:
# ============================================================
# CV PREDICTIVE PERFORMANCE
# Top-5, Top-8, Top-10, Top-13
# ============================================================

import pandas as pd
import numpy as np

from sklearn.model_selection import StratifiedKFold
from sklearn.impute import SimpleImputer
from sklearn.feature_selection import SelectKBest, f_classif
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)

from xgboost import XGBClassifier

from doda import DODASelector
from doda.adapters.sklearn_adapter import SklearnAdapter
from doda.knowledge import JSONProvider
from doda.fusion import RankFusion


# ============================================================
# CONFIGURATION
# ============================================================

TOP_K_VALUES = [5, 8, 10, 13]

N_SPLITS = 5
RANDOM_STATE = 42

WEIGHTS_FILE = "heartdisease_weights.json"


# ============================================================
# DATA
# ============================================================

# IMPORTANT:
# X and y should be the ORIGINAL dataset before:
# - imputation
# - scaling
# - feature selection
#
# Example:
#
# X = df.drop(columns=["target"])
# y = df["target"]


# ============================================================
# CROSS-VALIDATION
# ============================================================

cv = StratifiedKFold(
    n_splits=N_SPLITS,
    shuffle=True,
    random_state=RANDOM_STATE
)


# ============================================================
# STORAGE
# ============================================================

results = []


# ============================================================
# MODEL DEFINITIONS
# ============================================================

models = {

    "Logistic Regression": LogisticRegression(
        max_iter=2000,
        random_state=RANDOM_STATE
    ),

    "Random Forest": RandomForestClassifier(
        n_estimators=300,
        random_state=RANDOM_STATE,
        n_jobs=-1
    ),

    "XGBoost": XGBClassifier(
        n_estimators=300,
        max_depth=3,
        learning_rate=0.05,
        subsample=0.8,
        colsample_bytree=0.8,
        eval_metric="logloss",
        random_state=RANDOM_STATE,
        n_jobs=-1
    )
}


# ============================================================
# MAIN EXPERIMENT
# ============================================================

for method in ["LASSO", "DODA"]:

    print("\n")
    print("=" * 80)
    print(f"METHOD: {method}")
    print("=" * 80)

    for TOP_K in TOP_K_VALUES:

        print("\n" + "-" * 80)
        print(f"{method} | TOP-{TOP_K}")
        print("-" * 80)

        for model_name, base_model in models.items():

            print(f"\nModel: {model_name}")

            for fold, (train_idx, val_idx) in enumerate(
                cv.split(X, y),
                start=1
            ):

                # ====================================================
                # TRAIN / VALIDATION SPLIT
                # ====================================================

                X_train = X.iloc[train_idx].copy()
                X_val = X.iloc[val_idx].copy()

                y_train = y.iloc[train_idx].copy()
                y_val = y.iloc[val_idx].copy()


                # ====================================================
                # IMPUTATION
                # Fit ONLY on training data
                # ====================================================

                imputer = SimpleImputer(strategy="median")

                X_train_imp = pd.DataFrame(
                    imputer.fit_transform(X_train),
                    columns=X_train.columns,
                    index=X_train.index
                )

                X_val_imp = pd.DataFrame(
                    imputer.transform(X_val),
                    columns=X_val.columns,
                    index=X_val.index
                )


                # ====================================================
                # FEATURE SELECTION
                # ====================================================

                if method == "LASSO":

                    selector = SelectKBest(
                        score_func=f_classif,
                        k=TOP_K
                    )

                    X_train_selected = selector.fit_transform(
                        X_train_imp,
                        y_train
                    )

                    X_val_selected = selector.transform(
                        X_val_imp
                    )

                    selected_features = (
                        X_train_imp.columns[
                            selector.get_support()
                        ].tolist()
                    )


                elif method == "DODA":

                    # ------------------------------------------------
                    # LASSO provides mathematical feature scores
                    # DODA then incorporates clinical weights
                    # ------------------------------------------------

                    lasso_operator = SklearnAdapter(
                        SelectKBest(
                            score_func=f_classif,
                            k="all"
                        )
                    )

                    provider = JSONProvider(
                        WEIGHTS_FILE
                    )

                    doda_selector = DODASelector(
                        operators=[lasso_operator],
                        provider=provider,
                        fusion=RankFusion(),
                        top_k=TOP_K
                    )

                    doda_selector.fit(
                        X_train_imp,
                        y_train
                    )

                    selected_features = list(
                        doda_selector.get_selected_features()
                    )

                    X_train_selected = (
                        X_train_imp[selected_features]
                    )

                    X_val_selected = (
                        X_val_imp[selected_features]
                    )


                # ====================================================
                # MODEL
                # ====================================================

                # Create a fresh model for every fold
                model = type(base_model)(
                    **base_model.get_params()
                )


                # ====================================================
                # SCALING
                # Only Logistic Regression needs scaling
                # ====================================================

                if model_name == "Logistic Regression":

                    scaler = StandardScaler()

                    X_train_final = scaler.fit_transform(
                        X_train_selected
                    )

                    X_val_final = scaler.transform(
                        X_val_selected
                    )

                else:

                    X_train_final = X_train_selected
                    X_val_final = X_val_selected


                # ====================================================
                # TRAIN
                # ====================================================

                model.fit(
                    X_train_final,
                    y_train
                )


                # ====================================================
                # PREDICTION
                # ====================================================

                y_pred = model.predict(
                    X_val_final
                )

                y_prob = model.predict_proba(
                    X_val_final
                )[:, 1]


                # ====================================================
                # METRICS
                # ====================================================

                accuracy = accuracy_score(
                    y_val,
                    y_pred
                )

                precision = precision_score(
                    y_val,
                    y_pred,
                    zero_division=0
                )

                recall = recall_score(
                    y_val,
                    y_pred,
                    zero_division=0
                )

                f1 = f1_score(
                    y_val,
                    y_pred,
                    zero_division=0
                )

                auc = roc_auc_score(
                    y_val,
                    y_prob
                )


                # ====================================================
                # STORE RESULTS
                # ====================================================

                results.append({

                    "Method": method,

                    "Top_K": TOP_K,

                    "Model": model_name,

                    "Fold": fold,

                    "Accuracy": accuracy,

                    "Precision": precision,

                    "Recall": recall,

                    "F1": f1,

                    "ROC_AUC": auc,

                    "Selected_Features":
                        ", ".join(selected_features)
                })


                print(
                    f"  Fold {fold}: "
                    f"Acc={accuracy:.4f}, "
                    f"Prec={precision:.4f}, "
                    f"Recall={recall:.4f}, "
                    f"F1={f1:.4f}, "
                    f"AUC={auc:.4f}"
                )


# ============================================================
# RESULTS DATAFRAME
# ============================================================

results_df = pd.DataFrame(results)


# ============================================================
# SAVE FOLD-LEVEL RESULTS
# ============================================================

results_df.to_csv(
    "../../../results/heart_disease/lasso_cv_results.csv",
    index=False
)


# ============================================================
# SUMMARY — MEAN ± STD
# ============================================================

summary_df = (
    results_df
    .groupby(
        ["Method", "Top_K", "Model"]
    )
    .agg({

        "Accuracy": ["mean", "std"],

        "Precision": ["mean", "std"],

        "Recall": ["mean", "std"],

        "F1": ["mean", "std"],

        "ROC_AUC": ["mean", "std"]

    })
    .reset_index()
)


# ============================================================
# FLATTEN COLUMN NAMES
# ============================================================

summary_df.columns = [

    "Method",
    "Top_K",
    "Model",

    "Accuracy_Mean",
    "Accuracy_STD",

    "Precision_Mean",
    "Precision_STD",

    "Recall_Mean",
    "Recall_STD",

    "F1_Mean",
    "F1_STD",

    "ROC_AUC_Mean",
    "ROC_AUC_STD"
]


# ============================================================
# SAVE SUMMARY
# ============================================================

summary_df.to_csv(
    "../../../results/heart_disease/lasso_cv_summary.csv",
    index=False
)


# ============================================================
# DISPLAY SUMMARY
# ============================================================

print("\n")
print("=" * 100)
print("FINAL CV PERFORMANCE SUMMARY")
print("=" * 100)

display(
    summary_df.round(4)
)


# ============================================================
# BEST CONFIGURATION BY ROC-AUC
# ============================================================

best_auc = summary_df.loc[
    summary_df["ROC_AUC_Mean"].idxmax()
]

print("\n")
print("=" * 80)
print("BEST CONFIGURATION — ROC-AUC")
print("=" * 80)

print(
    f"Method:     {best_auc['Method']}"
)

print(
    f"Top-K:      {int(best_auc['Top_K'])}"
)

print(
    f"Model:      {best_auc['Model']}"
)

print(
    f"ROC-AUC:    "
    f"{best_auc['ROC_AUC_Mean']:.4f} "
    f"± {best_auc['ROC_AUC_STD']:.4f}"
)


# ============================================================
# BEST CONFIGURATION BY F1
# ============================================================

best_f1 = summary_df.loc[
    summary_df["F1_Mean"].idxmax()
]

print("\n")
print("=" * 80)
print("BEST CONFIGURATION — F1")
print("=" * 80)

print(
    f"Method:     {best_f1['Method']}"
)

print(
    f"Top-K:      {int(best_f1['Top_K'])}"
)

print(
    f"Model:      {best_f1['Model']}"
)

print(
    f"F1:         "
    f"{best_f1['F1_Mean']:.4f} "
    f"± {best_f1['F1_STD']:.4f}"
)



METHOD: LASSO

--------------------------------------------------------------------------------
LASSO | TOP-5
--------------------------------------------------------------------------------

Model: Logistic Regression


  Fold 1: Acc=0.8852, Prec=0.8621, Recall=0.8929, F1=0.8772, AUC=0.9313
  Fold 2: Acc=0.8197, Prec=0.8696, Recall=0.7143, F1=0.7843, AUC=0.8820
  Fold 3: Acc=0.8361, Prec=0.8462, Recall=0.7857, F1=0.8148, AUC=0.9156
  Fold 4: Acc=0.8167, Prec=0.8077, Recall=0.7778, F1=0.7925, AUC=0.8844
  Fold 5: Acc=0.8333, Prec=0.8750, Recall=0.7500, F1=0.8077, AUC=0.8571

Model: Random Forest
  Fold 1: Acc=0.7869, Prec=0.8000, Recall=0.7143, F1=0.7547, AUC=0.9129
  Fold 2: Acc=0.7705, Prec=0.7692, Recall=0.7143, F1=0.7407, AUC=0.8328
  Fold 3: Acc=0.8361, Prec=0.8000, Recall=0.8571, F1=0.8276, AUC=0.9080
  Fold 4: Acc=0.7333, Prec=0.6897, Recall=0.7407, F1=0.7143, AUC=0.8384
  Fold 5: Acc=0.7333, Prec=0.7500, Recall=0.6429, F1=0.6923, AUC=0.8432

Model: XGBoost
  Fold 1: Acc=0.8033, Prec=0.8077, Recall=0.7500, F1=0.7778, AUC=0.8999
  Fold 2: Acc=0.8197, Prec=0.8400, Recall=0.7500, F1=0.7925, AUC=0.8550
  Fold 3: Acc=0.8361, Prec=0.8214, Recall=0.8214, F1=0.8214, AUC=0.9080
  Fold 4: Acc=0.7167, Prec

,Method,Top_K,Model,Accuracy_Mean,Accuracy_STD,Precision_Mean,Precision_STD,Recall_Mean,Recall_STD,F1_Mean,F1_STD,ROC_AUC_Mean,ROC_AUC_STD
0,DODA,5,Logistic Regression,0.8315,0.0366,0.8444,0.0434,0.7767,0.0698,0.8077,0.0454,0.8954,0.0290
1,DODA,5,Random Forest,0.7886,0.0312,0.7890,0.0480,0.7410,0.0773,0.7616,0.0411,0.8707,0.0367
2,DODA,5,XGBoost,0.7919,0.0358,0.7949,0.0580,0.7410,0.0464,0.7657,0.0377,0.8723,0.0289
3,DODA,8,Logistic Regression,0.8349,0.0376,0.8512,0.0463,0.7767,0.0698,0.8108,0.0465,0.9004,0.0259
4,DODA,8,Random Forest,0.8316,0.0303,0.8499,0.0395,0.7690,0.0644,0.8062,0.0426,0.9040,0.0204
5,DODA,8,XGBoost,0.7852,0.0439,0.7754,0.0329,0.7476,0.1032,0.7588,0.0612,0.8809,0.0200
6,DODA,10,Logistic Regression,0.8415,0.0193,0.8509,0.0442,0.7984,0.0745,0.8210,0.0279,0.9099,0.0232
7,DODA,10,Random Forest,0.8317,0.0213,0.8362,0.0468,0.7910,0.0488,0.8114,0.0254,0.9085,0.0271
8,DODA,10,XGBoost,0.8184,0.0353,0.8186,0.0358,0.7765,0.0797,0.7953,0.0472,0.8902,0.0211
9,DODA,13,Logistic Regression,0.8316,0.0554,0.8304,0.0719,0.7984,0.0745,0.8126,0.0613,0.9116,0.0199




BEST CONFIGURATION — ROC-AUC
Method:     DODA
Top-K:      13
Model:      Logistic Regression
ROC-AUC:    0.9116 ± 0.0199


BEST CONFIGURATION — F1
Method:     DODA
Top-K:      13
Model:      Random Forest
F1:         0.8222 ± 0.0320


In [31]:
# =============================================================================
# REPEATED STRATIFIED CROSS-VALIDATION
# LASSO VS DODA-LASSO WITH RANK FUSION
# =============================================================================

import numpy as np
import pandas as pd

from sklearn.base import clone
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)

from sklearn.feature_selection import SelectFromModel

from doda import DODASelector
from doda.adapters.sklearn_adapter import SklearnAdapter
from doda.knowledge import JSONProvider
from doda.fusion import RankFusion


# =============================================================================
# CONFIGURATION
# =============================================================================

TOP_K_VALUES = [5, 8, 10, 13]

N_SPLITS = 5
N_REPEATS = 5
RANDOM_STATE = 42

target_column = "target"

X = df.drop(columns=[target_column])
y = df[target_column]

feature_names = X.columns.tolist()


# =============================================================================
# MODELS
# =============================================================================

models = {
    "Logistic Regression": LogisticRegression(
        max_iter=2000,
        random_state=RANDOM_STATE
    ),

    "Random Forest": RandomForestClassifier(
        n_estimators=300,
        random_state=RANDOM_STATE,
        n_jobs=-1
    ),

    "XGBoost": XGBClassifier(
        n_estimators=300,
        max_depth=3,
        learning_rate=0.05,
        subsample=0.8,
        colsample_bytree=0.8,
        eval_metric="logloss",
        random_state=RANDOM_STATE,
        n_jobs=-1
    )
}


# =============================================================================
# CLINICAL KNOWLEDGE PROVIDER
# =============================================================================

provider = JSONProvider(
    "heartdisease_weights.json"
)


# =============================================================================
# REPEATED STRATIFIED CROSS-VALIDATION
# =============================================================================

cv = RepeatedStratifiedKFold(
    n_splits=N_SPLITS,
    n_repeats=N_REPEATS,
    random_state=RANDOM_STATE
)


# =============================================================================
# RESULTS STORAGE
# =============================================================================

results = []


# =============================================================================
# CROSS-VALIDATION LOOP
# =============================================================================

for run_number, (train_idx, test_idx) in enumerate(
    cv.split(X, y),
    start=1
):

    print("\n" + "=" * 80)
    print(f"REPEATED STRATIFIED CV RUN {run_number}")
    print("=" * 80)

    X_train = X.iloc[train_idx].copy()
    X_test = X.iloc[test_idx].copy()

    y_train = y.iloc[train_idx].copy()
    y_test = y.iloc[test_idx].copy()


    # =========================================================================
    # IMPUTATION
    # =========================================================================

    imputer = SimpleImputer(
        strategy="median"
    )

    X_train_imp_array = imputer.fit_transform(
        X_train
    )

    X_test_imp_array = imputer.transform(
        X_test
    )

    X_train_imp = pd.DataFrame(
        X_train_imp_array,
        columns=feature_names,
        index=X_train.index
    )

    X_test_imp = pd.DataFrame(
        X_test_imp_array,
        columns=feature_names,
        index=X_test.index
    )


    # =========================================================================
    # FEATURE SELECTION FOR EACH TOP-K VALUE
    # =========================================================================

    for top_k in TOP_K_VALUES:

        print("\n" + "-" * 80)
        print(f"TOP-K = {top_k}")
        print("-" * 80)


        # =====================================================================
        # BASELINE LASSO FEATURE SELECTION
        # =====================================================================

        # Standardization is performed only using training data.
        # This is important because LASSO is scale-sensitive.

        lasso_scaler = StandardScaler()

        X_train_lasso_scaled = lasso_scaler.fit_transform(
            X_train_imp
        )

        X_test_lasso_scaled = lasso_scaler.transform(
            X_test_imp
        )

        lasso_model = LogisticRegression(
            penalty="l1",
            solver="liblinear",
            C=1.0,
            max_iter=2000,
            random_state=RANDOM_STATE
        )

        lasso_model.fit(
            X_train_lasso_scaled,
            y_train
        )

        lasso_coefficients = np.abs(
            lasso_model.coef_[0]
        )

        lasso_ranked_indices = np.argsort(
            lasso_coefficients
        )[::-1]

        lasso_selected_indices = (
            lasso_ranked_indices[:top_k]
        )

        lasso_features = [
            feature_names[i]
            for i in lasso_selected_indices
        ]

        print("LASSO-selected features:")
        print(lasso_features)


        # =====================================================================
        # DODA-LASSO FEATURE SELECTION
        # =====================================================================

        # DODA receives the L1 logistic-regression estimator directly.
        # Do not wrap it inside SelectFromModel because the DODA adapter
        # extracts the coef_ attribute from the estimator.

        doda_lasso_scaler = StandardScaler()

        X_train_doda_scaled_array = (
            doda_lasso_scaler.fit_transform(
                X_train_imp
            )
        )

        X_test_doda_scaled_array = (
            doda_lasso_scaler.transform(
                X_test_imp
            )
        )

        X_train_doda_scaled = pd.DataFrame(
            X_train_doda_scaled_array,
            columns=feature_names,
            index=X_train_imp.index
        )

        X_test_doda_scaled = pd.DataFrame(
            X_test_doda_scaled_array,
            columns=feature_names,
            index=X_test_imp.index
        )

        doda_lasso_estimator = LogisticRegression(
            penalty="l1",
            solver="liblinear",
            C=1.0,
            max_iter=2000,
            random_state=RANDOM_STATE
        )

        lasso_operator = SklearnAdapter(
            doda_lasso_estimator
        )

        doda_selector = DODASelector(
            operators=[lasso_operator],
            provider=provider,
            fusion=RankFusion(),
            top_k=top_k
        )

        doda_selector.fit(
            X_train_doda_scaled,
            y_train
        )

        doda_features = (
            doda_selector.get_selected_features()
        )

        print("DODA-selected features:")
        print(doda_features)


        # =====================================================================
        # PREPARE SELECTED FEATURE MATRICES
        # =====================================================================

        X_train_lasso_selected = (
            X_train_imp[lasso_features]
        )

        X_test_lasso_selected = (
            X_test_imp[lasso_features]
        )

        X_train_doda_selected = (
            X_train_imp[doda_features]
        )

        X_test_doda_selected = (
            X_test_imp[doda_features]
        )


        # =====================================================================
        # MODEL EVALUATION
        # =====================================================================

        selected_feature_sets = {
            "LASSO": (
                X_train_lasso_selected,
                X_test_lasso_selected,
                lasso_features
            ),

            "DODA": (
                X_train_doda_selected,
                X_test_doda_selected,
                doda_features
            )
        }


        for method, (
            X_train_selected,
            X_test_selected,
            selected_features
        ) in selected_feature_sets.items():

            for model_name, model_template in models.items():

                model = clone(
                    model_template
                )

                X_train_model = X_train_selected.copy()
                X_test_model = X_test_selected.copy()


                # -------------------------------------------------------------
                # Scale only Logistic Regression
                # -------------------------------------------------------------

                if model_name == "Logistic Regression":

                    model_scaler = StandardScaler()

                    X_train_model = (
                        model_scaler.fit_transform(
                            X_train_model
                        )
                    )

                    X_test_model = (
                        model_scaler.transform(
                            X_test_model
                        )
                    )


                # -------------------------------------------------------------
                # Train model
                # -------------------------------------------------------------

                model.fit(
                    X_train_model,
                    y_train
                )

                y_pred = model.predict(
                    X_test_model
                )

                y_prob = model.predict_proba(
                    X_test_model
                )[:, 1]


                # -------------------------------------------------------------
                # Calculate metrics
                # -------------------------------------------------------------

                accuracy = accuracy_score(
                    y_test,
                    y_pred
                )

                precision = precision_score(
                    y_test,
                    y_pred,
                    zero_division=0
                )

                recall = recall_score(
                    y_test,
                    y_pred,
                    zero_division=0
                )

                f1 = f1_score(
                    y_test,
                    y_pred,
                    zero_division=0
                )

                roc_auc = roc_auc_score(
                    y_test,
                    y_prob
                )


                # -------------------------------------------------------------
                # Store results
                # -------------------------------------------------------------

                results.append({
                    "Run": run_number,
                    "Method": method,
                    "Top_K": top_k,
                    "Model": model_name,
                    "Accuracy": accuracy,
                    "Precision": precision,
                    "Recall": recall,
                    "F1": f1,
                    "ROC_AUC": roc_auc,
                    "Selected_Features": selected_features
                })


# =============================================================================
# RESULTS DATAFRAME
# =============================================================================

results_df = pd.DataFrame(
    results
)

print("\n" + "=" * 80)
print("REPEATED STRATIFIED CV RESULTS")
print("=" * 80)

display(
    results_df.head()
)

print(
    "Total evaluation rows:",
    len(results_df)
)


REPEATED STRATIFIED CV RUN 1



--------------------------------------------------------------------------------
TOP-K = 5
--------------------------------------------------------------------------------
LASSO-selected features:
['ca', 'sex', 'exang', 'thal', 'cp']
Provider inside knowledge engine: <doda.knowledge.providers.json_provider.JSONProvider object at 0x00000207B38F6D50>
DODA Pipeline Started
Running: LogisticRegression
DODA feature order:
['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak']
Number of features: 13
Number of coefficients: 13
Operator Scores:
{'LogisticRegression_1': {'age': np.float64(0.07145485510224234), 'sex': np.float64(0.6323756123194068), 'cp': np.float64(0.39462315365466594), 'trestbps': np.float64(0.3892426381034713), 'chol': np.float64(0.15668728662651754), 'fbs': np.float64(0.2785872640017071), 'restecg': np.float64(0.15063212223381026), 'thalach': np.float64(0.3563576975002547), 'exang': np.float64(0.5652033929482854), 'oldpeak': np.float64(0.3

,Run,Method,Top_K,Model,Accuracy,Precision,Recall,F1,ROC_AUC,Selected_Features
0,1,LASSO,5,Logistic Regression,0.852459,0.851852,0.821429,0.836364,0.916126,"[ca, sex, exang, thal, cp]"
1,1,LASSO,5,Random Forest,0.786885,0.857143,0.642857,0.734694,0.873918,"[ca, sex, exang, thal, cp]"
2,1,LASSO,5,XGBoost,0.819672,0.869565,0.714286,0.784314,0.911797,"[ca, sex, exang, thal, cp]"
3,1,DODA,5,Logistic Regression,0.852459,0.851852,0.821429,0.836364,0.916126,"[ca, exang, thal, cp, sex]"
4,1,DODA,5,Random Forest,0.786885,0.857143,0.642857,0.734694,0.870671,"[ca, exang, thal, cp, sex]"


Total evaluation rows: 600


In [32]:
# =============================================================================
# PREPARE FEATURE SETS FOR STABILITY ANALYSIS
# =============================================================================

import ast
import numpy as np
import pandas as pd
from itertools import combinations


def convert_to_set(features):
    """
    Convert Selected_Features into a Python set.
    Handles lists, tuples, NumPy arrays, and strings.
    """

    if isinstance(features, str):

        # Handle string representation of a list
        if features.startswith("[") and features.endswith("]"):
            try:
                features = ast.literal_eval(features)
            except Exception:
                features = features.split(",")

        else:
            features = features.split(",")

    return {
        str(feature).strip().strip("'").strip('"')
        for feature in features
        if str(feature).strip()
    }


# Use only one model because feature selection is independent of evaluation model
stability_df = results_df[
    results_df["Model"] == "Logistic Regression"
].copy()

stability_df["Feature_Set"] = stability_df["Selected_Features"].apply(
    convert_to_set
)

print("Rows used for stability analysis:", len(stability_df))
print(stability_df[["Run", "Method", "Top_K", "Feature_Set"]].head())

Rows used for stability analysis: 200
    Run Method  Top_K                                        Feature_Set
0     1  LASSO      5                         {thal, ca, cp, sex, exang}
3     1   DODA      5                         {thal, ca, cp, sex, exang}
6     1  LASSO      8  {thal, ca, oldpeak, cp, sex, trestbps, thalach...
9     1   DODA      8  {thal, slope, ca, oldpeak, cp, sex, trestbps, ...
12    1  LASSO     10  {thal, slope, ca, fbs, oldpeak, cp, sex, trest...


In [33]:
# =============================================================================
# TEST 1: FEATURE SELECTION FREQUENCY
# =============================================================================

selection_frequency = []

for method in stability_df["Method"].unique():

    for top_k in sorted(stability_df["Top_K"].unique()):

        subset = stability_df[
            (stability_df["Method"] == method) &
            (stability_df["Top_K"] == top_k)
        ]

        total_runs = len(subset)

        all_features = sorted(
            set().union(*subset["Feature_Set"])
        )

        for feature in all_features:

            selected_count = sum(
                feature in feature_set
                for feature_set in subset["Feature_Set"]
            )

            selection_frequency.append({
                "Method": method,
                "Top_K": top_k,
                "Feature": feature,
                "Selected_Count": selected_count,
                "Total_Runs": total_runs,
                "Selection_Frequency_Percentage": (
                    selected_count / total_runs
                ) * 100
            })

selection_frequency_df = pd.DataFrame(selection_frequency)

selection_frequency_df = selection_frequency_df.sort_values(
    by=[
        "Method",
        "Top_K",
        "Selection_Frequency_Percentage"
    ],
    ascending=[True, True, False]
).reset_index(drop=True)

display(selection_frequency_df)

,Method,Top_K,Feature,Selected_Count,Total_Runs,Selection_Frequency_Percentage
0,DODA,5,ca,25,25,100.0
1,DODA,5,cp,24,25,96.0
2,DODA,5,thal,24,25,96.0
3,DODA,5,exang,18,25,72.0
4,DODA,5,sex,13,25,52.0
...,...,...,...,...,...,...
87,LASSO,13,sex,25,25,100.0
88,LASSO,13,slope,25,25,100.0
89,LASSO,13,thal,25,25,100.0
90,LASSO,13,thalach,25,25,100.0


In [34]:
# =============================================================================
# TOP FREQUENTLY SELECTED FEATURES
# =============================================================================

top_frequency_features = (
    selection_frequency_df
    .groupby(["Method", "Top_K"], group_keys=False)
    .head(10)
    .reset_index(drop=True)
)

display(top_frequency_features)

,Method,Top_K,Feature,Selected_Count,Total_Runs,Selection_Frequency_Percentage
0,DODA,5,ca,25,25,100.0
1,DODA,5,cp,24,25,96.0
2,DODA,5,thal,24,25,96.0
3,DODA,5,exang,18,25,72.0
4,DODA,5,sex,13,25,52.0
...,...,...,...,...,...,...
73,LASSO,13,fbs,25,25,100.0
74,LASSO,13,oldpeak,25,25,100.0
75,LASSO,13,restecg,25,25,100.0
76,LASSO,13,sex,25,25,100.0


In [35]:
# =============================================================================
# TEST 2: PAIRWISE JACCARD SIMILARITY
# =============================================================================

jaccard_results = []

for method in stability_df["Method"].unique():

    for top_k in sorted(stability_df["Top_K"].unique()):

        subset = stability_df[
            (stability_df["Method"] == method) &
            (stability_df["Top_K"] == top_k)
        ].reset_index(drop=True)

        feature_sets = subset["Feature_Set"].tolist()

        pairwise_scores = []

        for set_a, set_b in combinations(feature_sets, 2):

            intersection_size = len(set_a.intersection(set_b))
            union_size = len(set_a.union(set_b))

            if union_size == 0:
                jaccard_score = 1.0
            else:
                jaccard_score = intersection_size / union_size

            pairwise_scores.append(jaccard_score)

        jaccard_results.append({
            "Method": method,
            "Top_K": top_k,
            "Mean_Jaccard": np.mean(pairwise_scores),
            "Std_Jaccard": np.std(pairwise_scores, ddof=1),
            "Minimum_Jaccard": np.min(pairwise_scores),
            "Maximum_Jaccard": np.max(pairwise_scores)
        })

jaccard_df = pd.DataFrame(jaccard_results)

display(jaccard_df)

,Method,Top_K,Mean_Jaccard,Std_Jaccard,Minimum_Jaccard,Maximum_Jaccard
0,LASSO,5,0.670079,0.172081,0.250000,1.0
1,LASSO,8,0.752013,0.131222,0.454545,1.0
2,LASSO,10,0.766115,0.095308,0.538462,1.0
3,LASSO,13,1.000000,0.000000,1.000000,1.0
4,DODA,5,0.629841,0.176315,0.250000,1.0
5,DODA,8,0.782667,0.111722,0.600000,1.0
6,DODA,10,0.831313,0.094758,0.666667,1.0
7,DODA,13,1.000000,0.000000,1.000000,1.0


In [36]:
# =============================================================================
# TEST 3: KUNCHEVA INDEX
# =============================================================================

TOTAL_FEATURES = 13

kuncheva_results = []

for method in stability_df["Method"].unique():

    for top_k in sorted(stability_df["Top_K"].unique()):

        subset = stability_df[
            (stability_df["Method"] == method) &
            (stability_df["Top_K"] == top_k)
        ].reset_index(drop=True)

        feature_sets = subset["Feature_Set"].tolist()

        pairwise_scores = []

        for set_a, set_b in combinations(feature_sets, 2):

            k_a = len(set_a)
            k_b = len(set_b)

            # Skip pairs with different numbers of selected features
            if k_a != k_b:
                continue

            k = k_a

            # Kuncheva index is undefined for empty or complete feature sets
            if k == 0 or k >= TOTAL_FEATURES:
                continue

            common_features = len(set_a.intersection(set_b))

            expected_overlap = (k * k) / TOTAL_FEATURES

            denominator = k - expected_overlap

            if denominator == 0:
                continue

            kuncheva_score = (
                common_features - expected_overlap
            ) / denominator

            pairwise_scores.append(kuncheva_score)

        if len(pairwise_scores) == 0:

            kuncheva_results.append({
                "Method": method,
                "Top_K": top_k,
                "Mean_Kuncheva": np.nan,
                "Std_Kuncheva": np.nan,
                "Minimum_Kuncheva": np.nan,
                "Maximum_Kuncheva": np.nan,
                "Valid_Pair_Count": 0
            })

        else:

            kuncheva_results.append({
                "Method": method,
                "Top_K": top_k,
                "Mean_Kuncheva": np.mean(pairwise_scores),
                "Std_Kuncheva": (
                    np.std(pairwise_scores, ddof=1)
                    if len(pairwise_scores) > 1
                    else 0.0
                ),
                "Minimum_Kuncheva": np.min(pairwise_scores),
                "Maximum_Kuncheva": np.max(pairwise_scores),
                "Valid_Pair_Count": len(pairwise_scores)
            })

kuncheva_df = pd.DataFrame(kuncheva_results)

display(kuncheva_df)

,Method,Top_K,Mean_Kuncheva,Std_Kuncheva,Minimum_Kuncheva,Maximum_Kuncheva,Valid_Pair_Count
0,LASSO,5,0.658750,0.199132,0.025000,1.0,300
1,LASSO,8,0.615417,0.223115,0.025000,1.0,300
2,LASSO,10,0.412111,0.261441,-0.300000,1.0,300
3,LASSO,13,NaN,NaN,NaN,NaN,0
4,DODA,5,0.607833,0.216909,0.025000,1.0,300
5,DODA,8,0.671750,0.181225,0.350000,1.0,300
6,DODA,10,0.588333,0.243292,0.133333,1.0,300
7,DODA,13,NaN,NaN,NaN,NaN,0


In [37]:
# =============================================================================
# TEST 4: LASSO VS DODA FEATURE-SET AGREEMENT
# =============================================================================

agreement_results = []

for run in sorted(stability_df["Run"].unique()):

    for top_k in sorted(stability_df["Top_K"].unique()):

        lasso_rows = stability_df[
            (stability_df["Run"] == run) &
            (stability_df["Method"] == "LASSO") &
            (stability_df["Top_K"] == top_k)
        ]

        doda_rows = stability_df[
            (stability_df["Run"] == run) &
            (stability_df["Method"] == "DODA") &
            (stability_df["Top_K"] == top_k)
        ]

        if lasso_rows.empty or doda_rows.empty:
            continue

        lasso_set = lasso_rows.iloc[0]["Feature_Set"]
        doda_set = doda_rows.iloc[0]["Feature_Set"]

        intersection_size = len(lasso_set.intersection(doda_set))
        union_size = len(lasso_set.union(doda_set))

        if union_size == 0:
            jaccard_agreement = 1.0
        else:
            jaccard_agreement = intersection_size / union_size

        agreement_results.append({
            "Run": run,
            "Top_K": top_k,
            "LASSO_Features": ", ".join(sorted(lasso_set)),
            "DODA_Features": ", ".join(sorted(doda_set)),
            "Common_Feature_Count": intersection_size,
            "LASSO_Feature_Count": len(lasso_set),
            "DODA_Feature_Count": len(doda_set),
            "Same_Feature_Set": lasso_set == doda_set,
            "Jaccard_Agreement": jaccard_agreement
        })

agreement_df = pd.DataFrame(agreement_results)

display(agreement_df)

,Run,Top_K,LASSO_Features,DODA_Features,Common_Feature_Count,LASSO_Feature_Count,DODA_Feature_Count,Same_Feature_Set,Jaccard_Agreement
0,1,5,"ca, cp, exang, sex, thal","ca, cp, exang, sex, thal",5,5,5,True,1.000000
1,1,8,"ca, cp, exang, oldpeak, sex, thal, thalach, tr...","ca, cp, exang, oldpeak, sex, slope, thal, tres...",7,8,8,False,0.777778
2,1,10,"ca, cp, exang, fbs, oldpeak, sex, slope, thal,...","ca, chol, cp, exang, oldpeak, sex, slope, thal...",9,10,10,False,0.818182
3,1,13,"age, ca, chol, cp, exang, fbs, oldpeak, restec...","age, ca, chol, cp, exang, fbs, oldpeak, restec...",13,13,13,True,1.000000
4,2,5,"ca, cp, sex, thal, thalach","ca, cp, exang, sex, thal",4,5,5,False,0.666667
...,...,...,...,...,...,...,...,...,...
95,24,13,"age, ca, chol, cp, exang, fbs, oldpeak, restec...","age, ca, chol, cp, exang, fbs, oldpeak, restec...",13,13,13,True,1.000000
96,25,5,"ca, cp, sex, slope, thal","ca, cp, sex, slope, thal",5,5,5,True,1.000000
97,25,8,"ca, cp, exang, sex, slope, thal, thalach, tres...","ca, cp, exang, sex, slope, thal, thalach, tres...",8,8,8,True,1.000000
98,25,10,"ca, chol, cp, exang, fbs, sex, slope, thal, th...","ca, chol, cp, exang, oldpeak, sex, slope, thal...",9,10,10,False,0.818182


In [38]:
# =============================================================================
# SUMMARY OF LASSO VS DODA AGREEMENT
# =============================================================================

agreement_summary_df = (
    agreement_df
    .groupby("Top_K")
    .agg(
        Same_Feature_Set_Percentage=(
            "Same_Feature_Set",
            lambda x: x.mean() * 100
        ),
        Mean_Jaccard_Agreement=(
            "Jaccard_Agreement",
            "mean"
        ),
        Std_Jaccard_Agreement=(
            "Jaccard_Agreement",
            "std"
        ),
        Mean_Common_Features=(
            "Common_Feature_Count",
            "mean"
        )
    )
    .reset_index()
)

display(agreement_summary_df)

,Top_K,Same_Feature_Set_Percentage,Mean_Jaccard_Agreement,Std_Jaccard_Agreement,Mean_Common_Features
0,5,48.0,0.817143,0.185286,4.44
1,8,68.0,0.928889,0.105799,7.68
2,10,52.0,0.912727,0.092709,9.52
3,13,100.0,1.000000,0.000000,13.00


## LASSO Feature Selection — Heart Disease

### Overall Interpretation

The LASSO experiment evaluated conventional LASSO feature selection against **LASSO + Rank Fusion DODA**, using Logistic Regression, Random Forest, and XGBoost across Top-5, Top-8, Top-10, and Top-13 feature subsets.

Unlike the earlier Hadamard fusion approach, the current DODA implementation uses **Rank Fusion**, combining the LASSO-derived feature ranking with clinically informed feature ranking. The results show that Rank Fusion substantially changes the selected feature composition at lower Top-K values while retaining a clinically consistent core of features.

The results demonstrate that DODA does **not uniformly improve every predictive metric or every model**. However, it provides strong feature-selection stability and, particularly at Top-8 and Top-10, maintains or improves predictive performance for several models.

---

## 1. LASSO Baseline Feature Selection

The LASSO baseline identified a consistent core of heart-disease-related features.

At Top-5, the selected features were:

`ca, thal, cp, exang, sex`

At Top-8, the feature set expanded to:

`ca, thal, cp, exang, sex, thalach, slope, oldpeak`

At Top-10, `restecg` and `chol` were additionally included, while Top-13 retained all available features.

This indicates that LASSO consistently prioritised `ca`, `thal`, `cp`, and `exang` among the most important predictors.

---

## 2. Effect of Rank Fusion DODA on Feature Selection

After applying Rank Fusion DODA, the Top-5 feature set changed from:

**LASSO:**
`ca, thal, cp, exang, sex`

to:

**LASSO + DODA:**
`ca, thal, cp, exang, oldpeak`

Thus, DODA replaced `sex` with `oldpeak` at Top-5.

This is important because `oldpeak` was strongly retained by the DODA ranking: it was selected in **23/25 runs (92%)** at Top-5.

At Top-8, DODA selected:

`ca, thal, cp, exang, oldpeak, slope, thalach, age`

Here, `sex` was replaced by `age` compared with the LASSO baseline.

At Top-10, DODA selected:

`ca, thal, cp, exang, oldpeak, slope, thalach, sex, age, trestbps`

Compared with the LASSO Top-10 subset, DODA introduced `age` and `trestbps` while removing `restecg` and `chol`.

Therefore, Rank Fusion does not simply reproduce the LASSO ranking. It modifies the feature composition by incorporating the clinical ranking while preserving several of LASSO's strongest features.

---

## 3. Top-5 Predictive Performance

At Top-5, DODA produced mixed changes across the three models.

### Logistic Regression

DODA increased:

* Accuracy: **81.97% → 85.25%**
* Recall: **85.71% → 89.29%**
* F1-score: **81.36% → 84.75%**
* ROC-AUC: **0.8983 → 0.9405**

This represents an improvement across all reported metrics.

### Random Forest

DODA decreased accuracy from **81.97% to 72.13%** and F1-score from **80.70% to 72.13%**.

However, ROC-AUC decreased only slightly from **0.9048 to 0.8799**.

### XGBoost

DODA produced a small reduction in accuracy (**83.61% → 78.69%**) and F1-score (**82.76% → 77.19%**), while ROC-AUC decreased from **0.9242 to 0.8793**.

Therefore, at Top-5, the effect of DODA was strongly model-dependent. Logistic Regression benefited substantially, whereas the tree-based models performed better with the original LASSO feature subset.

---

## 4. Top-8 Predictive Performance

At Top-8, DODA selected:

`ca, thal, cp, exang, oldpeak, slope, thalach, sex`

The DODA subset produced:

### Logistic Regression

* Accuracy: **85.25% → 85.25%**
* F1-score: **84.75% → 84.75%**
* ROC-AUC: **0.9318 → 0.9351**

Thus, predictive performance was essentially preserved, with a small improvement in ROC-AUC.

### Random Forest

* Accuracy remained **83.61%**
* F1-score increased from **82.76% to 83.87%**
* ROC-AUC decreased slightly from **0.9432 to 0.9389**

This indicates broadly comparable performance.

### XGBoost

DODA increased:

* Accuracy: **81.97% → 85.25%**
* F1-score: **81.36% → 84.21%**
* ROC-AUC: **0.9307 → 0.9253**

The slight decrease in ROC-AUC occurred alongside improvements in accuracy and F1-score.

Overall, Top-8 provides evidence that DODA can alter the feature set while maintaining or improving several predictive measures.

---

## 5. Top-10 Predictive Performance

At Top-10, DODA selected:

`ca, thal, cp, exang, oldpeak, slope, thalach, sex, age, trestbps`

This produced particularly consistent results.

### Logistic Regression

* Accuracy: **85.25% → 86.89%**
* Recall: **92.86% → 92.86%**
* F1-score: **85.25% → 86.67%**
* ROC-AUC: **0.9329 → 0.9448**

Thus, DODA improved all major predictive measures except recall, which remained unchanged.

### Random Forest

* Accuracy: **88.52% → 86.89%**
* F1-score: **88.14% → 86.67%**
* ROC-AUC: **0.9416 → 0.9481**

Although accuracy and F1-score decreased, ROC-AUC improved.

### XGBoost

* Accuracy: **83.61% → 85.25%**
* Recall: **89.29% → 96.43%**
* F1-score: **83.33% → 85.71%**
* ROC-AUC: **0.9134 → 0.9416**

The most notable improvement at Top-10 was for XGBoost, particularly the increase in recall from **89.29% to 96.43%** and ROC-AUC from **0.9134 to 0.9416**.

---

## 6. Top-13

At Top-13, all available features are retained, so the feature-selection difference between LASSO and DODA disappears.

The two approaches therefore operate on the same complete feature set.

This provides a useful reference point because any differences at Top-5, Top-8, and Top-10 can be attributed to the different feature-selection strategies rather than different numbers of available predictors.

---

## 7. Cross-Validation Results

The repeated 5-fold results provide a more robust view than the single evaluation results.

For LASSO, mean ROC-AUC values ranged from:

* **0.8941–0.9116** for Logistic Regression
* **0.8671–0.9104** for Random Forest
* **0.8634–0.8931** for XGBoost

For LASSO + DODA, mean ROC-AUC values ranged from:

* **0.8954–0.9116** for Logistic Regression
* **0.8707–0.9085** for Random Forest
* **0.8723–0.8964** for XGBoost

The cross-validation results therefore show that DODA generally maintained **comparable discriminative performance** to LASSO rather than producing a universal performance increase.

The strongest DODA result in the repeated evaluation occurs at Top-10 for Logistic Regression, where ROC-AUC is **0.9099**, compared with **0.9099** for LASSO, while accuracy increases from **83.82% to 84.15%**.

For XGBoost at Top-10, DODA increases mean accuracy from **80.53% to 81.84%**, recall from **74.07% to 77.65%**, and F1-score from **77.70% to 79.53%**.

---

## 8. Feature-Selection Stability

Feature-selection stability is one of the strongest findings of the LASSO-DODA experiment.

| Top-K | Mean Jaccard |         SD | Interpretation      |
| ----: | -----------: | ---------: | ------------------- |
|     5 |       0.9011 |     0.1525 | High stability      |
|     8 |   **1.0000** | **0.0000** | Complete stability  |
|    10 |   **0.9473** |     0.0826 | Very high stability |
|    13 |   **1.0000** | **0.0000** | Complete stability  |

The Top-8 DODA feature set was selected identically across all 25 runs.

All eight features:

`ca, thal, cp, exang, oldpeak, slope, thalach, age`

had a selection frequency of **100%**.

At Top-10, most features were also highly stable:

* `ca`: 100%
* `thal`: 100%
* `cp`: 100%
* `exang`: 100%
* `oldpeak`: 100%
* `slope`: 100%
* `thalach`: 100%
* `age`: 100%
* `sex`: 96%
* `trestbps`: 88%

Only `restecg`, which appeared occasionally during the repeated Top-10 selections, had a low selection frequency of **16%**.

This indicates that the DODA ranking has a particularly stable core of eight features.

---

## 9. Overall Interpretation

The Heart Disease LASSO experiment demonstrates that **Rank Fusion DODA modifies the LASSO feature ranking while preserving a strong and highly reproducible core of features**.

The most consistent features across the DODA rankings were:

`ca, thal, cp, exang, oldpeak, slope, thalach, age`

The Top-8 configuration is particularly notable because all eight features were selected in **25/25 repeated runs**, producing a mean Jaccard similarity of **1.0000**.

Predictively, DODA does not consistently outperform LASSO across every model and Top-K configuration. Instead, its effect depends on the downstream model and number of selected features. Logistic Regression generally showed stable or improved performance, while Random Forest and XGBoost showed both improvements and reductions depending on Top-K.

The Top-10 configuration provides particularly useful evidence for DODA, with improvements in several classification metrics and a substantial increase in XGBoost recall. At the same time, the cross-validation results demonstrate that these improvements should be interpreted as **model- and Top-K-specific rather than as a universal performance gain**.

Overall, the LASSO experiment supports the role of Rank Fusion DODA as a **clinically guided re-ranking mechanism that can produce highly stable feature subsets while maintaining broadly comparable predictive performance**. Its contribution is therefore not limited to maximizing predictive accuracy, but also includes the integration of clinical information and the generation of reproducible feature rankings.
